src/python/pants/task/simple_codegen_task.py

```diff
@@ -248,24 +276,39 @@ def _inject_synthetic_target(
     :param target: The target to inject a synthetic target for.
     :param target_workdir: The work directory containing the generated code for the target.
     :param fingerprint: The fingerprint to create the synthetic target
-           with to avoid re-fingerprinting
-    :param mark_transitive_invalidation_hash_dirty: Whether to walk the build graph to invalidate
-           dependees of the target.
+           with to avoid re-fingerprinting.
     """
     copied_attributes = {}
     for attribute in self._copy_target_attributes:
       copied_attributes[attribute] = getattr(target, attribute)
 
     target_workdir = self.synthetic_target_dir(target, target_workdir)
 
+    synthetic_extra_dependencies = self.synthetic_target_extra_dependencies(target, target_workdir)
+    synthetic_target_type = self.synthetic_target_type(target)
+
+    if hasattr(synthetic_target_type, 'export_specs'):
+      original_exports = self._original_export_specs(target)
+      extra_exports = self.synthetic_target_extra_exports(target, target_workdir)
+
+      extra_exports_not_in_extra_dependencies = set(extra_exports).difference(
+        set(synthetic_extra_dependencies))
+      if len(extra_exports_not_in_extra_dependencies) > 0:
+        raise self.MismatchedExtraExports(
+          'Extra synthetic exports included targets not in the extra dependencies: {}'
```

Comment:
This should be validated much, much earlier, and raised as a (subclass of) TaskError. It indicates invalid configuration of particular options, and that is not obvious enough here.

Reply:
Oh. Hm. Do you have a good thought for how to do it? It looks a bit tricky because the extra_deps / exports take targets rather than target types.

The reason it's not a TaskError is because there might not be an end-user fix. The repo owner would need to fix it by changing configuration or possibly by changing the simple code gen extending task.

Reply:
Well, I don't feel strongly about the TaskError bit.

But it should happen much earlier... maybe as an explicit validation step at the beginning of `def execute`.

Reply:
Hm. So, iterate over all of the codegen targets and compute the extra deps / exports, then do it again during codegen? Or maybe I could memoize the calls.

Reply:
Hm, that's not what I meant, so I suspect I'm misreading this.

Reply:
I did the above, and it seems reasonable. None of the calls are marked as public, so as long as nothing uses them, it shouldn't break downstream users.

It also has the effect of preventing extra_dependencies from referencing synthetic targets generated during code gen. I think that is a constraint that makes sense.